File: dart/lib/contrast/contrast.dart
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:material_color_utilities/utils/color_utils.dart';
import 'package:material_color_utilities/utils/math_utils.dart';

/// Utility methods for calculating contrast given two colors, or calculating a
/// color given one color and a contrast ratio.
///
/// Contrast ratio is calculated using XYZ's Y. When linearized to match human
/// perception, Y becomes HCT's tone and L*a*b*'s' L*. Informally, this is the
/// lightness of a color.
///
/// Methods refer to tone, T in the the HCT color space.
/// Tone is equivalent to L* in the L*a*b* color space, or L in the LCH color
/// space.
class Contrast {
  /// Returns a contrast ratio, which ranges from 1 to 21.
  /// [toneA] Tone between 0 and 100. Values outside will be clamped.
  /// [toneB] Tone between 0 and 100. Values outside will be clamped.
  static double ratioOfTones(double toneA, double toneB) {
    toneA = MathUtils.clampDouble(0.0, 100.0, toneA);
    toneB = MathUtils.clampDouble(0.0, 100.0, toneB);
    return _ratioOfYs(
        ColorUtils.yFromLstar(toneA), ColorUtils.yFromLstar(toneB));
  }

  static double _ratioOfYs(double y1, double y2) {
    final lighter = y1 > y2 ? y1 : y2;
    final darker = (lighter == y2) ? y1 : y2;
    return (lighter + 5.0) / (darker + 5.0);
  }

  /// Returns a tone >= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns -1 if [ratio] cannot be achieved with [tone].
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in -1 being returned.
  /// [ratio] Contrast ratio of return value and [tone].
  /// Range is 1 to 21, invalid values have undefined behavior.
  static double lighter({required double tone, required double ratio}) {
    if (tone < 0.0 || tone > 100.0) {
      return -1.0;
    }

    final darkY = ColorUtils.yFromLstar(tone);
    final lightY = ratio * (darkY + 5.0) - 5.0;
    final realContrast = _ratioOfYs(lightY, darkY);
    final delta = (realContrast - ratio).abs();
    if (realContrast < ratio && delta > 0.04) {
      return -1;
    }

    // Ensure gamut mapping, which requires a 'range' on tone, will still result
    // the correct ratio by darkening slightly.
    final returnValue = ColorUtils.lstarFromY(lightY) + 0.4;
    if (returnValue < 0 || returnValue > 100) {
      return -1;
    }
    return returnValue;
  }

  /// Returns a tone <= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns -1 if [ratio] cannot be achieved with [tone].
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in -1 being returned.
  /// [ratio] Contrast ratio of return value and [tone].
  /// Range is 1 to 21, invalid values have undefined behavior.
  static double darker({required double tone, required double ratio}) {
    if (tone < 0.0 || tone > 100.0) {
      return -1.0;
    }

    final lightY = ColorUtils.yFromLstar(tone);
    final darkY = ((lightY + 5.0) / ratio) - 5.0;
    final realContrast = _ratioOfYs(lightY, darkY);

    final delta = (realContrast - ratio).abs();
    if (realContrast < ratio && delta > 0.04) {
      return -1;
    }

    // Ensure gamut mapping, which requires a 'range' on tone, will still result
    // the correct ratio by darkening slightly.
    final returnValue = ColorUtils.lstarFromY(darkY) - 0.4;
    if (returnValue < 0 || returnValue > 100) {
      return -1;
    }
    return returnValue;
  }

  /// Returns a tone >= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns 100 if [ratio] cannot be achieved with [tone].
  ///
  /// This method is unsafe because the returned value is guaranteed to be in
  /// bounds for tone, i.e. between 0 and 100. However, that value may not reach
  /// the [ratio] with [tone]. For example, there is no color lighter than T100.
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in 100 being returned.
  /// [ratio] Desired contrast ratio of return value and tone parameter.
  /// Range is 1 to 21, invalid values have undefined behavior.
  static double lighterUnsafe({required double tone, required double ratio}) {
    final lighterSafe = lighter(tone: (tone), ratio: ratio);
    return (lighterSafe < 0.0) ? 100.0 : lighterSafe;
  }

  /// Returns a tone <= [tone] that ensures [ratio].
  /// Return value is between 0 and 100.
  /// Returns 0 if [ratio] cannot be achieved with [tone].
  ///
  /// This method is unsafe because the returned value is guaranteed to be in
  /// bounds for tone, i.e. between 0 and 100. However, that value may not reach
  /// the [ratio] with [tone]. For example, there is no color darker than T0.
  ///
  /// [tone] Tone return value must contrast with.
  /// Range is 0 to 100. Invalid values will result in 0 being returned.
  /// [ratio] Desired contrast ratio of return value and tone parameter.
  /// Range is 1 to 21, invalid values have undefined behavior.
  static double darkerUnsafe({required double tone, required double ratio}) {
    final darkerSafe = darker(tone: (tone), ratio: ratio);
    return (darkerSafe < 0.0) ? 0.0 : darkerSafe;
  }
}

File: cpp/contrast/contrast.cc
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/contrast/contrast.h"

#include <algorithm>
#include <cmath>

#include "cpp/utils/utils.h"

namespace material_color_utilities {
// Given a color and a contrast ratio to reach, the luminance of a color that
// reaches that ratio with the color can be calculated. However, that luminance
// may not contrast as desired, i.e. the contrast ratio of the input color
// and the returned luminance may not reach the contrast ratio  asked for.
//
// When the desired contrast ratio and the result contrast ratio differ by
// more than this amount,  an error value should be returned, or the method
// should be documented as 'unsafe', meaning, it will return a valid luminance
// but that luminance may not meet the requested contrast ratio.
//
// 0.04 selected because it ensures the resulting ratio rounds to the
// same tenth.
constexpr double CONTRAST_RATIO_EPSILON = 0.04;

// Color spaces that measure luminance, such as Y in XYZ, L* in L*a*b*,
// or T in HCT, are known as  perceptual accurate color spaces.
//
// To be displayed, they must gamut map to a "display space", one that has
// a defined limit on the number of colors. Display spaces include sRGB,
// more commonly understood as RGB/HSL/HSV/HSB.
//
// Gamut mapping is undefined and not defined by the color space. Any
// gamut mapping algorithm must choose how to sacrifice accuracy in hue,
// saturation, and/or lightness.
//
// A principled solution is to maintain lightness, thus maintaining
// contrast/a11y, maintain hue, thus maintaining aesthetic intent, and reduce
// chroma until the color is in gamut.
//
// HCT chooses this solution, but, that doesn't mean it will _exactly_ matched
// desired lightness, if only because RGB is quantized: RGB is expressed as
// a set of integers: there may be an RGB color with, for example,
// 47.892 lightness, but not 47.891.
//
// To allow for this inherent incompatibility between perceptually accurate
// color spaces and display color spaces, methods that take a contrast ratio
// and luminance, and return a luminance that reaches that contrast ratio for
// the input luminance, purposefully darken/lighten their result such that
// the desired contrast ratio will be reached even if inaccuracy is introduced.
//
// 0.4 is generous, ex. HCT requires much less delta. It was chosen because
// it provides a rough guarantee that as long as a percetual color space
// gamut maps lightness such that the resulting lightness rounds to the same
// as the requested, the desired contrast ratio will be reached.
constexpr double LUMINANCE_GAMUT_MAP_TOLERANCE = 0.4;

double RatioOfYs(double y1, double y2) {
  double lighter = y1 > y2 ? y1 : y2;
  double darker = (lighter == y2) ? y1 : y2;
  return (lighter + 5.0) / (darker + 5.0);
}

double RatioOfTones(double tone_a, double tone_b) {
  tone_a = std::clamp(tone_a, 0.0, 100.0);
  tone_b = std::clamp(tone_b, 0.0, 100.0);
  return RatioOfYs(YFromLstar(tone_a), YFromLstar(tone_b));
}

double Lighter(double tone, double ratio) {
  if (tone < 0.0 || tone > 100.0) {
    return -1.0;
  }

  double dark_y = YFromLstar(tone);
  double light_y = ratio * (dark_y + 5.0) - 5.0;
  double real_contrast = RatioOfYs(light_y, dark_y);
  double delta = abs(real_contrast - ratio);
  if (real_contrast < ratio && delta > CONTRAST_RATIO_EPSILON) {
    return -1;
  }

  // ensure gamut mapping, which requires a 'range' on tone, will still result
  // the correct ratio by darkening slightly.
  double value = LstarFromY(light_y) + LUMINANCE_GAMUT_MAP_TOLERANCE;
  if (value < 0 || value > 100) {
    return -1;
  }
  return value;
}

double Darker(double tone, double ratio) {
  if (tone < 0.0 || tone > 100.0) {
    return -1.0;
  }

  double light_y = YFromLstar(tone);
  double dark_y = ((light_y + 5.0) / ratio) - 5.0;
  double real_contrast = RatioOfYs(light_y, dark_y);

  double delta = abs(real_contrast - ratio);
  if (real_contrast < ratio && delta > CONTRAST_RATIO_EPSILON) {
    return -1;
  }

  // ensure gamut mapping, which requires a 'range' on tone, will still result
  // the correct ratio by darkening slightly.
  double value = LstarFromY(dark_y) - LUMINANCE_GAMUT_MAP_TOLERANCE;
  if (value < 0 || value > 100) {
    return -1;
  }
  return value;
}

double LighterUnsafe(double tone, double ratio) {
  double lighter_safe = Lighter(tone, ratio);
  return (lighter_safe < 0.0) ? 100.0 : lighter_safe;
}

double DarkerUnsafe(double tone, double ratio) {
  double darker_safe = Darker(tone, ratio);
  return (darker_safe < 0.0) ? 0.0 : darker_safe;
}

}  // namespace material_color_utilities
